scripts/gis-material-update/process/modules/hsl.py

```diff
@@ -1,14 +1,381 @@
+            hours += start_times
+
+        hour_range = []
+        for h_s, m_s, h_e, m_e in hours:
+            hour_range.append(
+                ("n_{:02d}{:02d}".format(h_s, m_s), time(h_s, m_s), time(h_e, m_e))
+            )
+        return hour_range
+
+    def _compute_rush_hours(self, stops_and_trips: pd.DataFrame) -> pd.DataFrame:
+        """Compute rush hours for given data frame."""
+        rush_hours = self._rush_hours()
+
+        # construct list of tuples(<descriptor>, <partial function>)
+        # to help computing multiple columns to data frame
+        pl = [(rh[0], partial(self._agg_rush_hour, rh[1], rh[2])) for rh in rush_hours]
+
+        kwargs = {nm: lambda x: pf(x.departure_time) for nm, pf in pl}
+        return stops_and_trips.assign(**kwargs)
+
+    def _parse_time(self, t_candidate: str) -> dict:
+        """Parse time. Support hour values larger than 23.
+        If hour part is larger than 23, "next_day" boolean value is True"""
+        format_string = "{:d}:{:d}:{:d}"
+        h, m, s = parse(format_string, t_candidate)
+        next_day = False
+        if h > 23:
+            h %= 24
+            next_day = True
+        time_str = "{:02d}:{:02d}:{:02d}".format(h, m, s)
+        t = datetime.strptime(time_str, "%H:%M:%S").time()
+        return {"time": t, "next_day": next_day}
+
+    def _agg_rush_hour(self, t_start: time, t_end: time, t_candidate):
+        """Check if time is between an interval. Return 1 if it is, 0 otherwise.
+
+        Handle both pd.Series and single value input for t_candidate."""
+
+        def is_between(candidate, low, high) -> bool:
+            """Check if candidate is in range [low, high)"""
+            return low <= candidate < high
+
+        if isinstance(t_candidate, pd.Series):
+            # Time in GTFS can be greater than 23.
+            # Hours falling to next day are not treated here
+            # due to it is extremely unlikely that day of operation
+            # continues until morning rush hour.
+
+            # TODO: handle previous day and current day for times past midnight.
+            t_c = t_candidate.apply(lambda r: self._parse_time(r).get("time"))
+            retval = 1 * t_c.apply(lambda r: is_between(r, t_start, t_end))
+            return retval
+        else:
+            # TODO: handle previous day and current day for times past midnight.
+            t_c = self._parse_time(t_candidate).get("time")
+            if is_between(t_c, t_start, t_end):
+                return 1
+            else:
+                return 0
+
+    def _route_is_trunk(self, route_id: str) -> str:
+        """Check if route is trunk route."""
+        is_trunk_route = {
+            r"^1500.*": "yes",
+            r"^2200.*": "yes",
+            r"^2510.*": "yes",
+            r"^2550.*": "yes",
+            r"^4560.*": "yes",
+            r"^1018.*": "almost",
+            r"^1039.*": "almost",
+            r"^1040.*": "almost",
```

Comment:
These are proper trunk lines these days. 18 Has been replaced by 20 and 39 by 30. There's also 570. https://www.hsl.fi/hsl/runkoverkko

If I remember correctly, the trunk lines had a different `route_type` then other HSL buses. This would be a more robust way to check for trunk lines than a list of ever-changing route ids.

Reply:
To be done in https://helsinkisolutionoffice.atlassian.net/browse/HAI-1327